datasource.go

```diff
@@ -133,38 +170,148 @@ func (ds *sqldatasource) QueryData(ctx context.Context, req *backend.QueryDataRe
+		return "", dbConnection{}, ErrorMissingDBConnection
 	}
-	if !ds.EnableMultipleConnections || len(q.ConnectionArgs) == 0 {
+	if !ds.EnableMultipleConnections || len(connArgs) == 0 {
 		return key, dbConn, nil
 	}
 
-	key = keyWithConnectionArgs(datasourceUID, q.ConnectionArgs)
+	key = keyWithConnectionArgs(datasourceUID, connArgs)
 	if cachedConn, ok := ds.getDBConnection(key); ok {
 		return key, cachedConn, nil
 	}
 
-	var err error
-	db, err := ds.c.Connect(dbConn.settings, q.ConnectionArgs)
+	db, err := ds.c.Connect(dbConn.settings, connArgs)
 	if err != nil {
 		return "", dbConnection{}, err
 	}
+
+	var asyncDB AsyncDB
+	if ds.asyncDriver != nil {
+		asyncDB, err = ds.asyncDriver.GetAsyncDB(dbConn.settings, connArgs)
+		if err != nil {
+			return "", dbConnection{}, err
+		}
+	}
+
 	// Assign this connection in the cache
-	dbConn = dbConnection{db, dbConn.settings}
+	dbConn = dbConnection{db: db, asyncDB: asyncDB, settings: dbConn.settings}
 	ds.storeDBConnection(key, dbConn)
 
 	return key, dbConn, nil
 }
 
+// handleQuery will call query, and attempt to reconnect if the query failed
+func (ds *sqldatasource) handleAsyncQuery(ctx context.Context, req backend.DataQuery, datasourceUID string) (data.Frames, error) {
+	// Convert the backend.DataQuery into a Query object
+	q, err := GetQuery(req)
+	if err != nil {
+		return getErrorFrameFromQuery(q), err
+	}
+
+	// Apply supported macros to the query
+	q.RawSQL, err = Interpolate(ds.c, q)
+	if err != nil {
+		return getErrorFrameFromQuery(q), fmt.Errorf("%s: %w", "Could not apply macros", err)
+	}
+
+	// Apply the default FillMode, overwritting it if the query specifies it
+	fillMode := ds.driverSettings.FillMode
+	if q.FillMissing != nil {
+		fillMode = q.FillMissing
+	}
+
+	// Retrieve the database connection
+	cacheKey, dbConn, err := ds.getDBConnectionFromConnArgs(datasourceUID, q.ConnectionArgs)
+	if err != nil {
+		return getErrorFrameFromQuery(q), err
+	}
+	if dbConn.asyncDB == nil {
+		return nil, fmt.Errorf("unable to get AsyncDB")
+	}
+
+	if ds.driverSettings.Timeout != 0 {
+		tctx, cancel := context.WithTimeout(ctx, ds.driverSettings.Timeout)
+		defer cancel()
+		ctx = tctx
+	}
+
+	if q.QueryID == "" {
+		queryID, err := startQuery(ctx, dbConn.asyncDB, q)
+		if err != nil {
+			return getErrorFrameFromQuery(q), err
+		}
+		return data.Frames{
+			{Meta: &data.FrameMeta{
+				ExecutedQueryString: q.RawSQL,
+				Custom:              queryMeta{QueryID: queryID, Status: "started"}},
+			},
+		}, nil
+	}
+
+	status, err := queryStatus(ctx, dbConn.asyncDB, q)
+	if err != nil {
+		return getErrorFrameFromQuery(q), err
+	}
+	customMeta := queryMeta{QueryID: q.QueryID, Status: status.String()}
+	if !status.Finished() {
+		return data.Frames{
+			{Meta: &data.FrameMeta{
+				ExecutedQueryString: q.RawSQL,
+				Custom:              customMeta},
+			},
+		}, nil
+	}
+
+	res, err := queryAsync(ctx, dbConn.db, ds.c.Converters(), fillMode, q)
+	if err == nil || errors.Is(err, ErrorNoResults) {
```

Comment:
I think `OR` is correct.

We want to treat the absence of an error as well as an "no results" error as the same since "no results" isn't really an error in our case.

Reply:
yeah, it's possible that the we get "no results" if the datasource hasn't found anything, but we don't want to retry on that or treat it like a connection error.